added alert manager section for user defined project rules (#270)
bcgov/platform-developer-docs: src/docs/app-monitoring/user-defined-monitoring.md

# Set up a User Defined Monitoring in OpenShift

> Last updated: **January 5, 2023**

Author: incorporated feedback and sent to review

**Data is currently stored for 15 days.**

## Monitor user-defined projects using Alertmanager

> Alerting rules are created for user-defined namespace and based on specific chosen metrics these alerts get triggered. You can create alerts based on the metrics you created above, or on the existing cluster metrics like pod memory usage.

Author: incorporated feedback and sent to review

> Create a YAML file lets assume the name is app-alerting-rule.yaml and we can add rule named `example-alert`, the rule triggers an alert when the `version` metric exposed by the sample service becomes `0`.

Author: incorporated feedback and sent to review

```yaml
apiVersion: monitoring.coreos.com/v1
kind: PrometheusRule
metadata:
  name: example-alert
  namespace: license-dev
spec:
  groups:
  - name: example
    rules:
    - alert: VersionAlert 
      for: 1m 
      expr: version{job="prometheus-example-app"} == 0 
      labels:
        severity: warning 
      annotations:
        message: This is an example alert. 
```

> Here we can see the following :

Author: incorporated feedback and sent to review

> - ``` alert: VersionAlert ``` which is the name of the alerting rule.

Author: incorporated feedback and sent to review

> - ``` for: 1m ``` is the duration before the alert is triggered if the condition is TRUE.

Author: incorporated feedback and sent to review

- ``` expr: version{job="prometheus-example-app"} == 0 ``` PromQL query expression that defines the new rule

> - ``` severity: warning ``` Severity level assigned to the alert to ensure users understand the impact and cause of the alert and it also lets us know who will get sent the alert and any other custom labels you like can also be included in the alert email.

Author: incorporated feedback and sent to review

> - ``` message: This is an example alert. ``` The message due to which the alert is triggered.

Author: incorporated feedback and sent to review

> Ensure you have the correct namespace for your alert, the code given above is a sample.

Reviewer: ```suggestion
Make sure your alert uses the correct namespace; the code provided is just an example.
```

Author: incorporated feedback and sent to review

> Once the alert rule is created the configuration file has to be applied to the cluster with the command given below.

Reviewer: ```suggestion
Once you've created the alert rule, apply the configuration file to the cluster using the following command:
```

Author: incorporated feedback and sent to review

```bash
oc apply -f app-alerting-rule.yaml
```

> `AlertmanagerConfig` is part of the Prometheus Operator, which is a Kubernetes operator that helps manage Prometheus instances and configurations in a Kubernetes environment. By using AlertmanagerConfig, you can keep your alerting rules and configurations within Kubernetes resources, making it easier to manage and version control them alongside your application code and infrastructure. For more information please refer to this [link](https://developer.gov.bc.ca/docs/default/component/platform-developer-docs/docs/platform-automation/alertmanager/)

Reviewer: ```suggestion
`AlertmanagerConfig` is part of the Prometheus Operator, which helps manage Prometheus instances and configurations in Kubernetes. Using AlertmanagerConfig ,allows you to manage and version control your alerting rules and configurations within Kubernetes resources, keeping them aligned with your application code and infrastructure. More information about [Alertmanager](https://developer.gov.bc.ca/docs/default/component/platform-developer-docs/docs/platform-automation/alertmanager/)
```

Author: incorporated feedback and sent to review

## Sysdig Monitor

> You can now let Sysdig agent collecting your custom metrics and show them in the Sysdig console. Add `prometheus.io/scrape=true` annotation set in your pod. The Sysdig agent will scrape your application pod and send its `/metrics` to the Sysdig console.

Reviewer: ```suggestion
You can now have the Sysdig agent collect your custom metrics and display them in the Sysdig console. Add the `prometheus.io/scrape=true` annotation to your pod. The Sysdig agent will then scrape your application pod and send its `/metrics` to the Sysdig console.
```

Author: incorporated feedback and sent to review

> ## Monitor user-defined projects using Alertmanager

Reviewer 2: Can we move this section before **Sysdig Monitor** ? To keep the on-cluster prometheus stuff and Sysdig stuff separate.

Reviewer 3: agree! if a team just need alerts but not a full dashboard, this is sufficient.

Author: moved the section before sysdig section

> Alerting rules are created for user-defined projects and based on specific chosen metrics these alerts get triggered.

Reviewer 2: I don't think people reading this will be thinking about the difference between infra and user projects. We also try and keep the `Project` keyword out of things as it has a special meaning in BCGov. `namespace` is better.

Reviewer 2: Can we add more language to tie this to the previous section that created the custom metric? Can we also explain that you can use the non-custom metrics that the platform already has for your alerts?

Reviewer 2: Can we add some info about the `AlertmanagerConfig` object that gets created for teams in their namespaces? Maybe also link to https://developer.gov.bc.ca/docs/default/component/platform-developer-docs/docs/platform-automation/alertmanager/

Author: converted project to namespace in doc and added language to tie up previous section.

> Create a YAML file lets assume the name is app-alerting-rule.yaml and we can add rule named ```example-alert```, the rule triggers an alert when the ```version``` metric exposed by the sample service becomes ```0```.

Reviewer 2: Inline code only needs a single backtick not three

Reviewer 2: I don't think we need to be explicit about creating yaml files and applying them. If you look at other sections of the doc like **Expose the metrics from your app** we just provide the yaml without details on how to create the object. It's assumed that skill exists.

> ```console

Reviewer 2: ```suggestion
```console
```

This is yaml code, not console output

> ```console

Reviewer 3: ```suggestion
```yaml
```

apiVersion: monitoring.coreos.com/v1
kind: PrometheusRule
metadata:
  name: example-alert

>   namespace: ns1

Reviewer 2: ```suggestion
  namespace: license-dev
```
Platform specific placeholder for a namespace

Author: made the change

spec:
  groups:
  - name: example
    rules:
    - alert: VersionAlert 
      for: 1m 
      expr: version{job="prometheus-example-app"} == 0 
      labels:
        severity: warning 
      annotations:
        message: This is an example alert. 
```

Here we can see the following :

- ``` alert: VersionAlert ``` which is the name of the alerting rule.
- ``` for: 1m ``` is the duration before the alert is triggered if the condition is TRUE.
- ``` expr: version{job="prometheus-example-app"} == 0 ``` PromQL query expression that defines the new rule

> - ``` severity: warning ``` Severity level assigned to the alert to ensure users understand the impact and cause of the alert

Reviewer 2: Severity also dictates who will get sent the alert (when in a prod namespace)
You can also add any other custom labels you'd like which will be included in the alert email

Author: added this portion.

- ``` message: This is an example alert. ``` The message due to which the alert is triggered.

Ensure you have the correct namespace for your alert, the code given above is a sample.

Once the alert rule is created the configuration file has to be applied to the cluster with the command given below.

```bash
oc apply -f app-alerting-rule.yaml
```

## Related pages

> 

Reviewer 2: Let's update these doc links from 4.10 to their equivalent page in the 4.14 docs. Maybe add a link to info on creating PrometheusRules?

Author: added https://docs.openshift.com/container-platform/4.14/observability/monitoring/managing-alerts.html#creating-alerting-rules-for-user-defined-projects_managing-alerts
